haystack/evaluation/eval.py

```diff
@@ -27,6 +34,85 @@ def __init__(
         self.outputs = outputs
         self.expected_outputs = expected_outputs
 
+        # Determine the type of the runnable
+        if str(type(runnable).__name__) == "Pipeline":
+            self.runnable_type = "pipeline"
+        else:
+            self.runnable_type = "component"
+
+    def calculate_metrics(self, metric: Union[Metric, Callable[..., MetricsResult]], **kwargs) -> MetricsResult:
+        """
+        Calculate evaluation metrics based on the provided Metric or using the custom metric function.
+
+        :param metric: The Metric indicating the type of metric to calculate or custom function to compute.
+        :return: MetricsResult containing the calculated metric.
+        """
+        if metric == Metric.EM:
+            predictions = get_answers_from_output(self.outputs, self.runnable_type)
+            labels = get_answers_from_output(self.expected_outputs, self.runnable_type)
+            return self._calculate_em(predictions=predictions, labels=labels, **kwargs)
+
+        return metric(self, **kwargs)
+
+    def _calculate_em(
+        self,
+        predictions,
+        labels,
+        regexes_to_ignore=None,
+        ignore_case=False,
+        ignore_punctuation=False,
+        ignore_numbers=False,
+    ):
+        """
+        Calculates the Exact Match (EM) score between two lists of predictions and labels.
+        Exact Match (EM) score measures the percentage of samples where the predicted text exactly matches the
+          corresponding ground truth label.
+
+        :param predictions: A list of predicted text strings.
+        :param labels (list): A list of ground truth (reference) text strings.
+        :param regexes_to_ignore (list, optional): A list of regular expressions. If provided, it removes substrings
+            matching these regular expressions from both predictions and labels before comparison. Defaults to None.
+        :param ignore_case (bool, optional): If True, performs case-insensitive comparison. Defaults to False.
+        :param ignore_punctuation (bool, optional): If True, removes punctuation from both predictions and labels before
+            comparison. Defaults to False.
+        :param ignore_numbers (bool, optional): If True, removes numerical digits from both predictions and labels
+            before comparison. Defaults to False.
+
+        :return: A MetricsResult object containing the calculated Exact Match (EM) score.
+        """
+
+        if len(predictions) != len(labels):
+            raise ValueError("The number of predictions and labels must be the same.")
+        if len(predictions) == len(labels) == 0:
+            # Return Exact Match as 0 for no inputs
+            return MetricsResult({"exact_match": 0.0})
+
+        if regexes_to_ignore is not None:
```

Comment:
All the preprocessing with `regexes_to_ignore`, `ignore_case`, `ignore_punctuation`, and `ignore_numbers` can be useful for other metrics too. We should extract that functionality into a separate function so that it can be reused in other, future `_calculate_metric` functions.

Reply:
I have created a `preprocess_text` function with this functionality and moved it to `eval_utils.py`. Unit tests have also been added.